### PR TITLE
use tokio mpsc so we can use blocking_send

### PR DESCRIPTION
**Pull request by CompuIves**
<a href="https://codesandbox.io/">CodeSandbox logoCodeSandbox logo</a>&nbsp; Open in CodeSandbox <a href="https://codesandbox.io/p/github/codesandbox/directory-permission-watcher/feat/blocking-send">Web Editor</a> | <a href="https://codesandbox.io/p/vscode?owner=codesandbox&repo=directory-permission-watcher&branch=feat/blocking-send">VS Code</a>

This will use tokio `mpsc` so we can use a thing called `blocking_send`. This will not block the thread and allow to continue watching for events. Also upped the backpressure of the channel from 1 to 512, might also help with perf.

### CODE REVIEW

**Comment by CompuIves** [comment]
I'm thinking we might want to remove `async` altogether in this case, since I don't think we use multiple threads in this case (no `tokio::spawn` going on). We could maybe go with an implementation where we have a single thread handling the messages, that way we don't have context switching and it might reduce the `futex` calls.

**Comment by CompuIves** [comment]
Added another commit where we use tokio multithreading, one thread is the watcher, the other thread is updating the permissions. We _could_ spawn an async thread for every permission update if we want, but that might be too much overhead for a simple operation.

File: src/chmod.rs [hunk 1]
```diff
@@ -1,23 +1,19 @@
...
-pub fn update_permission_recursive(path: PathBuf) {
-    if let Ok(path_str) = path.clone().into_os_string().into_string() {
-        let cmd_output = Command::new("chmod")
-            .args(["-R", "g+rw", path_str.as_str()])
-            .output();
+pub fn update_permission_recursive(path: &Path) {
+    let path_str = path.as_os_str();
+    let cmd_output = Command::new("chmod")
+        .args(["-R", "g+rw"])
+        .arg(path_str)
```
Comment (CompuIves) [inline src/chmod.rs:8]: Using this, we can pass `os_str` directly. So no need of conversion needed

File: src/main.rs [hunk 2]
```diff
@@ -1,92 +1,77 @@
...
     // Automatically select the best implementation for your platform.
     // You can also access each implementation directly e.g. INotifyWatcher.
     let watcher = RecommendedWatcher::new(move |res| {
-        futures::executor::block_on(async {
-            match tx.send(res).await {
-                Ok(()) => {}
-                Err(err) => println!("watch error: {:?}", err),
-            };
-        })
+        match tx.blocking_send(res) {
```
Comment (CompuIves) [inline src/main.rs:15]: With tokio we can do a blocking_send directly, not exactly sure how it works under the hood but I think it doesn't block unless needed by backpressure (and doesn't require a thread spawn).

File: src/main.rs [hunk 3]
```diff
@@ -1,92 +1,77 @@
...
     // below will be monitored for changes.
     watcher.watch(path.as_ref(), RecursiveMode::Recursive)?;
 
-    thread::sleep(time::Duration::from_millis(500));
+    tokio::time::sleep(time::Duration::from_millis(500)).await;
+    chmod::update_permission_recursive(path);
 
-    chmod::update_permission_recursive(path.clone());
+    // Spawn the watcher in a new thread so it can't block the main thread.
+    let thread = tokio::spawn(async move {
```
Comment (CompuIves) [inline src/main.rs:35]: Starting the permission handler in a separate thread from the watcher.

File: src/main.rs [hunk 4]
```diff
@@ -1,92 +1,77 @@
...
-                    permissions::check_permissions(paths);
                 }
+                Err(err) => println!("watch error: {:?}", err),
             }
-            Err(err) => println!("watch error: {:?}", err),
         }
-    }
+    });
+
+    thread.await.ok();
```
Comment (DeMoorJasper) [inline src/main.rs]: Shouldn't we handle this error? Or at least log it

**Review by DeMoorJasper — APPROVED** [review]

**Comment by DeMoorJasper** [comment]
Also pretty sure this didn't block the watcher as the watcher and permission fixer were always on a separate thread. Increasing the bus size and thread usage might also not be the greatest idea, I remember reducing thread usage because otherwise this would use too much ressources

File: src/main.rs (hunk 4 above)
Comment (CompuIves) [reply src/main.rs]: Yep, good point. This error would be thrown if the thread panics, I forgot to log it.

**Comment by CompuIves** [comment]
I _think_ that the current version only uses one thread, because `futures::block_on` will run the future on the _current_ thread, and I don't see any other `std::thread::spawn` or `task::spawn`. I agree that 2 threads should probably be the max, one thread for the watcher and the other one for the permission fixer.

**Comment by CompuIves** [comment]
But I think we can also get rid of tokio if we want to, since we only run two threads (with no meaningful `await` calls in those threads) we don't benefit a lot from tokio and it probably adds a bit of complexity and performance overhead. I wonder what you think about this @DeMoorJasper.

**Comment by DeMoorJasper** [comment]
My only concern was the error, everything else is fine. Pretty sure the watcher spawns a thread as well, so that was the other thread I was referring to, but this should be fine

**Pull request merged**